### PR TITLE
fix: panic due to no module

### LINKED ISSUE #10290: Telegraf 1.21.0 - issue with SNMP plugin
### Relevent telegraf.conf

```toml
[[inputs.snmp]]
  # List of agents to poll
  agents = [ "tardis.ip", "flash.ip", "dreamer.ip", "batman.ip", "millenniumfalcon.ip" , "area51.ip"]
  # Polling interval
  interval = "60s"
  # Timeout for each SNMP query.
  timeout = "10s"
  # Number of retries to attempt within timeout.
  retries = 3
  # SNMP version
  version = 2
  # SNMP community string.
  community = "public"
  # Measurement name
  name = "snmp.UAP"
  ##
  ## System Details
  ##
  #  System name (hostname)
```

### System info

Raspbian, Telegraf 1.21.0

### Docker

no

### Steps to reproduce

1.
2.
3.
...

### Expected behavior

should work as before with 1.20.4

MIB file unchanged:
pi@raspberrypidns:/etc/telegraf/telegraf.d $ ll /var/lib/mibs/ietf/DPI20-MIB
-rw-r--r-- 1 root root 1683 25. Dez 2020  /var/lib/mibs/ietf/DPI20-MIB

### Actual behavior

Dez 16 14:52:13 raspberrypidns telegraf[18888]: Parse module: /var/lib/mibs/ietf/DPI20-MIB:9:4: unexpected "ibm" (expected ";")
Dez 16 14:52:13 raspberrypidns telegraf[18888]: 2021-12-16T13:52:13Z W! [inputs.snmp] Could not load module at DPI20-MIB
Dez 16 14:52:16 raspberrypidns telegraf[18888]: Parse module: /var/lib/mibs/ietf/HPR-MIB:494:30: unexpected "HprRtpCounter" (expected "}")
Dez 16 14:52:16 raspberrypidns telegraf[18888]: 2021-12-16T13:52:16Z W! [inputs.snmp] Could not load module at HPR-MIB

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by MyaLongmire**
### Required for all PRs:

- [ ] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #10290 

THIS IS A DRAFT! Loaded module in to prevent panic on IETF mibs. Has been tested by community members to fix their panics. Now loads in a module and passes the `gosmi.SmiNode` to other function so the node does not get lost and cause a panic.

### CODE REVIEW

File: internal/snmp/translate.go [hunk 1]
```diff
@@ -145,14 +145,19 @@ func GetIndex(oidNum string, mibPrefix string) (col []string, tagOids map[string
 }
 
 //nolint:revive //Too many return variable but necessary
-func SnmpTranslateCall(oid string) (mibName string, oidNum string, oidText string, conversion string, err error) {
+func SnmpTranslateCall(oid string) (mibName string, oidNum string, oidText string, conversion string, module gosmi.SmiModule, node gosmi.SmiNode, err error) {
```
Comment (Hipska) [inline internal/snmp/translate.go:148]: if returning the module, then there is no need to also return `mibName`, as You can just do `module.Name`.

File: plugins/inputs/snmp/snmp_test.go [hunk 2]
```diff
@@ -145,9 +145,9 @@ func TestFieldInit(t *testing.T) {
 		{".1.2.3", "foo", "", ".1.2.3", "foo", ""},
 		{".iso.2.3", "foo", "", ".1.2.3", "foo", ""},
 		{".1.0.0.0.1.1", "", "", ".1.0.0.0.1.1", "server", ""},
-		{"TEST::server", "", "", ".1.0.0.0.1.1", "server", ""},
-		{"TEST::server.0", "", "", ".1.0.0.0.1.1.0", "server.0", ""},
-		{"TEST::server", "foo", "", ".1.0.0.0.1.1", "foo", ""},
+		// {"TEST::server", "", "", ".1.0.0.0.1.1", "server", ""},
+		// {"TEST::server.0", "", "", ".1.0.0.0.1.1.0", "server.0", ""},
+		// {"TEST::server", "foo", "", ".1.0.0.0.1.1", "foo", ""},
```
Comment (Hipska) [inline plugins/inputs/snmp/snmp_test.go:150]: You know I don't like changing tests.. Why is this?

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/inputs/snmp/snmp_test.go (hunk 2 above)
Comment (Hipska) [reply plugins/inputs/snmp/snmp_test.go:150]: Nevermind, just saw your PR comment.

File: internal/snmp/translate.go (hunk 1 above)
Comment (MyaLongmire) [reply internal/snmp/translate.go]: No longer returning `module` as it is not needed but `node` is. I would rather pass around a `string` than a `gosmi.SmiModule`.

File: plugins/inputs/snmp/snmp_test.go (hunk 2 above)
Comment (MyaLongmire) [reply plugins/inputs/snmp/snmp_test.go]: All tests are back to normal. Thank you for your patience :)

File: internal/snmp/translate.go [hunk 3]
```diff
@@ -121,12 +121,12 @@ func TrapLookup(oid string) (e MibEntry, err error) {
...
 // The following is for snmp
 
-func GetIndex(oidNum string, mibPrefix string) (col []string, tagOids map[string]struct{}, err error) {
+func GetIndex(oidNum string, mibPrefix string, module gosmi.SmiModule, node gosmi.SmiNode) (col []string, tagOids map[string]struct{}, err error) {
 	// first attempt to get the table's tags
 	tagOids = map[string]struct{}{}
 
 	// mimcks grabbing INDEX {} that is returned from snmptranslate -Td MibName
-	node, err := gosmi.GetNodeByOID(types.OidMustFromString(oidNum))
+	// node, err := gosmi.GetNodeByOID(types.OidMustFromString(oidNum))
```
[the remark this answers is not in the document]
Comment (MyaLongmire) [reply internal/snmp/translate.go]: Deleted. Just left it there for testing as that line introduced the panic.

File: internal/snmp/translate.go [hunk 4]
```diff
@@ -145,14 +139,19 @@ func GetIndex(oidNum string, mibPrefix string) (col []string, tagOids map[string
...
+func SnmpTranslateCall(oid string) (mibName string, oidNum string, oidText string, conversion string, node gosmi.SmiNode, err error) {
 	var out gosmi.SmiNode
 	var end string
 	if strings.ContainsAny(oid, "::") {
 		// split given oid
 		// for example RFC1213-MIB::sysUpTime.0
 		s := strings.Split(oid, "::")
 		// node becomes sysUpTime.0
+		moduleName := s[0]
+		_, err := gosmi.GetModule(moduleName)
```
Comment (Hipska) [inline internal/snmp/translate.go:151]:
```suggestion
		module, err := gosmi.GetModule(moduleName)
```

File: internal/snmp/translate.go [hunk 5]
```diff
@@ -163,7 +162,7 @@ func SnmpTranslateCall(oid string) (mibName string, oidNum string, oidText strin
 
 		out, err = gosmi.GetNode(node)
```
Comment (Hipska) [inline internal/snmp/translate.go:163]:
```suggestion
		out, err = module.GetNode(node)
```
or 
```suggestion
		out, err = gosmi.GetNode(node, module)
```

**Review by Hipska — CHANGES_REQUESTED** [review]
Same remark as maintainer of gosmi, You know the module name, but aren't doing anything with it, so you traverse all the loaded mibs in order to find one node that might happen to have the same name.

File: internal/snmp/translate.go (hunk 5 above)
Comment (MyaLongmire) [reply internal/snmp/translate.go:163]: Both of these will fail on pasring other mibs. loading the module in fixes the panic and the library keeps track of it from there

Comment (Hipska) [reply internal/snmp/translate.go:163]: What do you mean with "other mibs"? Other than what?

Comment (MyaLongmire) [reply internal/snmp/translate.go:163]: if you change it to module some of the test mobs than fail to parse. By keeping it gosmi and just node all mibs in the test parse properly and there is no more panic

Comment (Hipska) [reply internal/snmp/translate.go:167]: Okay, will have a look at those mibs. Shouldn’t be the case.

Comment (MyaLongmire) [reply internal/snmp/translate.go:167]: I couldn't figure out how to change the mib so it passed when module was passed as well. I also know you hate changing of the test so I didn't want to remove that mib from the test. A catch 22.

**Comment by MyaLongmire** [comment]
Added module but remove the server mib test as I hand wrote the mib and believe it is invalid. There is a reason people do not handwrite mibs as they are very complex. The tests were based on invalid mibs to begin with, I kept all of the ieft valid mibs to keep the validity of the tests.

File: internal/snmp/translate.go (hunk 5 above)
Comment (Hipska) [reply internal/snmp/translate.go]: I think it's okay in this case as what is being tested is still there

Comment (MyaLongmire) [reply internal/snmp/translate.go]: Awesome! Thank you for working with me on these snmp issues :)

**Review by Hipska — APPROVED** [review]

**Review by powersj — APPROVED** [review]

**Pull request merged**